### PR TITLE
fix: pending variable declarations

### PR DESCRIPTION
**Pull request by AirQuick**
In response to https://github.com/xspec/xspec/pull/1415#discussion_r602930299, this pull request changes how `x:variable` and `x:param[parent::x:scenario]` in a pending scenario are handled:
- They produce no variables, if they are considered as pending.
- They are not considered as pending, if they are inherited by a focused scenario.

### CODE REVIEW

File: src/compiler/base/main.xsl [hunk 1]
```diff
@@ -43,9 +43,11 @@
       Accumulators for scenario-level variable declarations (x:param and x:variable)
    -->
 
-   <!-- Push and pop variable declaration elements based on node identity -->
+   <!-- Push and pop variable declaration elements based on node identity. Pending variable
+      declaration elements are considered as if they did not exist. -->
    <xsl:accumulator name="stacked-vardecls" as="element()*" initial-value="()">
-      <xsl:accumulator-rule match="x:scenario/x:param | x:scenario/x:variable"
+      <xsl:accumulator-rule
+         match="(x:scenario/x:param | x:scenario/x:variable)[x:reason-for-pending(.) => empty()]"
```
Comment (galtm) [inline src/compiler/base/main.xsl:50]: Is the following interpretation correct? The reason for not including `[x:reason-for-pending(.) => empty()]` on line 64, in addition to here, is that it doesn't really matter on line 64. A pending variable or parameter definition will have the `stacked-vardecls-distinct-uqnames` accumulator data attached, but you will never need to retrieve it from that node.

File: test/external_unfocused-param_not-inherited-by-focus.xspec [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+	xmlns:x="http://www.jenitennison.com/xslt/xspec">
+
+	<!--
+		Test //x:scenario/x:param when they are unfocused by @focus on a non-inner scenario.
+	-->
+
+	<x:scenario focus="focus" label="Focused scenario" />
+
+	<x:scenario label="Unfocused scenario">
+		<x:param name="unfocused-scenario-param" select="error() div 0 + $bogus:nonexistent" />
```
Comment (galtm) [inline test/external_unfocused-param_not-inherited-by-focus.xspec:12]: The high density of deliberate problems with this XPath expression is amusing.

**Review by galtm — APPROVED** [review]
Is it worthwhile to have tests similar to the ones you have, except that the variables and parameters are defined inside shared scenarios and referenced using `x:like`? It would show that the inheritance logic accounts for resolution of `x:like` instead of being based only on the original XSpec document's tree structure. If you'd like me to create those variations (in a separate PR because I don't think I know how to add files to someone else's PR), let me know.

File: src/compiler/base/main.xsl (hunk 1 above)
Comment (AirQuick) [reply src/compiler/base/main.xsl:50]: 02bd7c66f59b371e59cd60adc5dd0a24cae4d39b add some comments.

**Comment by AirQuick** [comment]
> Is it worthwhile to have tests similar to the ones you have, except that the variables and parameters are defined inside shared scenarios and referenced using `x:like`? It would show that the inheritance logic accounts for resolution of `x:like` instead of being based only on the original XSpec document's tree structure. If you'd like me to create those variations (in a separate PR because I don't think I know how to add files to someone else's PR), let me know.

Good point. I didn't think about that. Can you create them?
I think, if you prefer, you can push commits directly to this PR by cloning its branch. i.e. `git clone -b discard-pending-vardecls https://github.com/AirQuick/xspec.git`. But a separate PR would be easier to handle.

**Comment by galtm** [comment]
> a separate PR would be easier to handle

Sure, I'll do that.

**Review by galtm — APPROVED** [review]

**Pull request merged**